test/octree/test_octree.cpp

```diff
@@ -1001,6 +1001,13 @@ TEST (PCL, Octree_Pointcloud_Iterator_Test)
   ASSERT_EQ (octreeA.getLeafCount (), leafNodeCount);
   ASSERT_EQ (octreeA.getBranchCount (), branchNodeCount);
 
+  // Test the operator==() and operator!=()
+  OctreePointCloud<PointXYZ>::BreadthFirstAtDepthIterator itA_end = octreeA.breadth_at_depth_end ();
+  OctreePointCloud<PointXYZ>::BreadthFirstAtDepthIterator itB_end = octreeA.breadth_at_depth_end ();
+
+  ASSERT_TRUE (itA_end == itB_end);
+  ASSERT_FALSE (itA_end != itB_end);
+  ASSERT_TRUE ((itA_end == itB_end) == !(itA_end != itB_end));
```

Comment:
The behavior I want to make sure that is also implemented is the following.
```cpp
OctreePointCloud<PointXYZ>::BreadthFirstAtDepthIterator itA_end = octreeA.breadth_at_depth_end ();
OctreePointCloud<PointXYZ>::BreadthFirstAtDepthIterator itB_end = octreeB.breadth_at_depth_end ();

ASSERT_FALSE (itA_end == itB_end);
ASSERT_TRUE (itA_end != itB_end);
ASSERT_TRUE ((itA_end == itB_end) == !(itA_end != itB_end));
```

Notice the iterators come from different octrees.

Reply:
Ok, I see. I do another check with octreeB. On the `ASSERT_FALSE (itA_end == itB_end)`, when I check to code with a debugger, this see this (file octree/include/pcl/octree/octree_iterator.h: l. 137):
```c++
        /** \brief Equal comparison operator
         * \param[in] other OctreeIteratorBase to compare with
         */
        bool operator==(const OctreeIteratorBase& other) const
        {
          return (( octree_ ==other.octree_) &&
                  ( current_state_ == other.current_state_) &&
                  ( max_octree_depth_ == other.max_octree_depth_) );
        }
```

In the return, all addresses are equal to `NULL`, even for the field `octree_`: that is the trouble!

Reply:
I try to fix this issue and I think maybe the better thing to do is declare sentinel octree (or maybe a sentinel state for the iterator) where the iterator should point to I the end. This way, you should not be annoying by deferrencing things 🙂

Reply:
Let me grasp your thoughts on what I originally devised:
- To make octree_ a required initialization parameter on OctreeIteratorBase, which means that any properly initialized iterator will have a pointer to it's octree object
- The end state, to be marked as `current_state_ = NULL`, still  retaining the value of it's octree.

Then we'll need to update all definitions for end iterators that exist in the current Octree classes, and also the behavior of incrementing from the last valid node to the "end" state.

I think this should be enough to provide proper discriminatory characteristics to the iterators to meet the equality/inequality requirements we have. Am I overlooking anything?

Reply:
You explain it better but we have the same idea. This should work, hopefully. I began some work on my repository: 
https://github.com/frozar/pcl/tree/octree_iterator_fix

Not everything is good, but at least you can see the files I modified.